server/internal/router/route.go

```diff
@@ -0,0 +1,132 @@
+package router
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+type route struct {
+	// m maps a model to a list of engine IPs.
+	m map[model][]string
+	// engines is the collections of engine IPs.
+	engines map[string]struct{}
+
+	// mu protects m and engines.
+	mu sync.Mutex
+}
+
+func newRoute() *route {
+	return &route{
+		m:       make(map[model][]string),
+		engines: make(map[string]struct{}),
+	}
+}
+
+func (r *route) getRoute(model model) []string {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.m[model]
+}
+
+func (r *route) addRoute(model model, ip string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	found, ok := r.m[model]
+	if !ok {
+		r.m[model] = []string{ip}
+		return
+	}
+	for _, v := range found {
+		if v == ip {
+			return
+		}
+	}
+	r.m[model] = append(found, ip)
+}
+
+func (r *route) deleteRoute(model model) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	for k := range r.m {
+		if k.id == model.id {
+			delete(r.m, k)
+			return nil
+		}
+	}
+	return fmt.Errorf("model %s not found", model)
+}
+
+func (r *route) addServer(ip string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.engines[ip] = struct{}{}
+}
+
+func (r *route) deleteServer(ip string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	for k, ips := range r.m {
+		for i, v := range ips {
+			if v == ip {
+				r.m[k] = append(ips[:i], ips[i+1:]...)
+				continue
+			}
+		}
+	}
+
+	delete(r.engines, ip)
+}
+
+func (r *route) replace(nr *route) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.m = nr.m
+	r.engines = nr.engines
+}
+
+// findLeastLoadedEngine finds the engine with the least number of loaded models.
+func (r *route) findLeastLoadedEngine() (string, error) {
+	modelsByIP := make(map[string][]model)
+
+	r.mu.Lock()
+	tmp := r.m
+	for ip := range r.engines {
+		modelsByIP[ip] = []model{}
+	}
+	r.mu.Unlock()
+
+	for k, v := range tmp {
+		for _, ip := range v {
+			modelsByIP[ip] = append(modelsByIP[ip], k)
+		}
+	}
+
+	min := 0
+	ip := ""
+	for k, v := range modelsByIP {
+		if len(v) < min || ip == "" {
+			min = len(v)
+			ip = k
+		}
+	}
+
+	if ip == "" {
+		return "", fmt.Errorf("no engine available")
+	}
+	return ip, nil
+}
+
+func (r *route) printRoute() {
```

Comment:
```suggestion
func (r *route) print() {
```

or `String()` and then let the caller print.